### PR TITLE
WooCommerce: Add variation modal.

### LINKED ISSUE #13512: WooCommerce: Add/Edit Product: Variations Modal
Next to each variation (#12770) should be a gear icon that opens an 'advanced details' modal for the variation. 

The modal should include a SKU component, a rich text editing area for the variation description, and a toggle for variation visibility.

### PR DESCRIPTION
**Pull request by justinshreve**
This adds the variation modal that displays when you click the settings cog for a variation in the variation UI (#13698).

It allows you to configure the description and visible status. It also contains a placeholder for the SKU which we can implement once #13653 is handled.

You can also switch between variations within the modal.

Closes #13512.

To test:
* Visit `http://calypso.localhost:3000/store/products/:site/add`
* Start typing in some variation types.
* Click the cog for a variation and see that the modal opens.
* Type in the description box and toggle visible status.
* Switch to another modal using the modal navigation.
* Click done, and click the cog for another variation, to make sure that variation opens instead.

screen shot 2017-05-09 at 9 13 38 am

@kellychoffman @jameskoster This PR contains UI elements for review.

### CODE REVIEW

File: client/extensions/woocommerce/app/products/product-form-variations-modal.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		const navVariations = ( variations && variations.filter( v => v.attributes.length > 0 ) ) || [];
+		const navItems = navVariations.map( function( v, i ) {
+			return (
+				<ModalNavItem key={ i } variation={ v } selected={ selectedVariation } />
+			);
+		} );
+
+		return (
+			<div className="products__product-form-modal-wrapper">
+				<VerticalMenu onClick={ this.switchVariation } className="products__product-form-modal-menu">
```
Comment (allendav) [inline client/extensions/woocommerce/app/products/product-form-variations-modal.js]: I'm curious - why don't we have to bind `this` to switchVariation?

**Comment by allendav** [comment]
At 980px wide, variation dimension input widths are inconsistent

screen shot 2017-05-09 at 12 37 39 pm

cc @kellychoffman @jameskoster

File: client/extensions/woocommerce/app/products/product-form-variations-modal.js (hunk 1 above)
Comment (justinshreve) [inline client/extensions/woocommerce/app/products/product-form-variations-modal.js]: We do: https://github.com/Automattic/wp-calypso/pull/13827/files/52cae7654981d26c3bb9383b52bd36e7ca21e480#diff-1d2b47198ed0be90a536ba1cff662b45R34

**Comment by kellychoffman** [comment]
> At 980px wide, variation dimension input widths are inconsistent

This is addressed in another PR.

File: client/extensions/woocommerce/app/products/product-form-variations-modal.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+					</FormFieldSet>
+
+					<FormLabel>
+						{ translate( 'Visible' ) }
+						<FormToggle
+							onChange={ this.toggleVisible }
+							checked={ variation.visible }
+						/>
+					</FormLabel>
+					<p>{ translate(
```
Comment (allendav) [inline client/extensions/woocommerce/app/products/product-form-variations-modal.js]: How about using `FormSettingExplanation ` ?

File: client/extensions/woocommerce/app/products/product-form-variations-modal.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+				</VerticalMenu>
+
+				<div className="products__product-form-modal-contents">
+					<h1>{ formattedVariationName( variation ) }</h1>
+					<span className="products__product-form-sku">SKU:</span>
+
+					<FormFieldSet className="products__product-form-variation-description">
+						<FormLabel htmlFor="description">{ translate( 'Description' ) }</FormLabel>
+						<FormTextArea name="description" value={ variation.description || '' } onChange={ this.setDescription } />
+						<p>{ translate(
```
Comment (allendav) [inline client/extensions/woocommerce/app/products/product-form-variations-modal.js]: How about using `FormSettingExplanation` ?

**Comment by allendav** [comment]
The cogs seem to be retaining focus state or something - open the modal by clicking on one, then close the modal and notice the slight difference in shade is retained:

screen shot 2017-05-09 at 1 48 20 pm

cc @kellychoffman @jameskoster

**Comment by kellychoffman** [comment]
@allendav: We are removing the cog in favor of linking up the variation name.

**Comment by justinshreve** [comment]
@allendav I have updated those strings to use `FormSettingExplanation` (thanks, I didn't know about it before). I have also made updates that remove the cog and link the name per the discussion with Kelly. I've also rebased this, so it should be good for you to continue reviewing.

**Comment by allendav** [comment]
Maybe in a later, separate issue, but I think we are going to need debouncing or something. This starts to get a little slow:

screen shot 2017-05-09 at 3 12 46 pm

**Comment by allendav** [comment]
If you end up with more than about a dozen variations, the modal vertical list cuts off the rest of them:

screen shot 2017-05-09 at 3 15 10 pm

File: client/extensions/woocommerce/app/products/product-form-variations-modal.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	switchVariation( selectedVariation ) {
+		this.setState( {
+			selectedVariation
+		} );
+	}
+
+	render() {
+		const { variations, translate } = this.props;
+		const { selectedVariation } = this.state;
+		const variation = find( variations, ( v ) => selectedVariation === v.id );
```
Comment (allendav) [inline client/extensions/woocommerce/app/products/product-form-variations-modal.js]: Maybe a helper function for finding the currently selected variation? You've used this code 3 places.

File: client/extensions/woocommerce/app/products/product-form-variations-modal.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+					<FormLabel>
+						{ translate( 'Visible' ) }
+						<FormToggle
+							onChange={ this.toggleVisible }
+							checked={ variation.visible }
+						/>
+					</FormLabel>
+					<FormSettingExplanation>{ translate(
+						'Hidden variations cannot be selected for purchase by customers.'
```
Comment (allendav) [inline client/extensions/woocommerce/app/products/product-form-variations-modal.js]: I wonder if we should only display this warning-like text if they toggle visible to false? cc @kellychoffman - when I see it right now I take an extra moment to think about it and make sure the toggle seems to be enabled

**Comment by kellychoffman** [comment]
👍 Styled. The only thing I can't figure out is how to make the modal 600px wide.

**Comment by jameskoster** [comment]
> The only thing I can't figure out is how to make the modal 600px wide.

It's because of [this](https://github.com/Automattic/wp-calypso/pull/13604#pullrequestreview-36855287). The modal exists outside of the `.woocommerce` wrapper so all css will just be ignored.

**Comment by justinshreve** [comment]
@allendav I have added a helper function for getting the selected variation, and I have added a vertical scroll bar for the menu when there are lots of variations to list. We discussed hiding that help text in Slack, but are going to leave as-is. I am handling debounce in a separate PR since it doesn't relate to these changes. Mind taking another look? Thanks!

**Comment by allendav** [comment]
@justinshreve - the modal is coming out way too tall - should probably be limited to the viewport height less some margin? cc @kellychoffman @jameskoster 

screen shot 2017-05-10 at 8 30 54 am

**Comment by jameskoster** [comment]
The variation menu needs a max-height and scrollbar I think.

**Comment by justinshreve** [comment]
It does now.

**Comment by allendav** [comment]
That works - ty!  There is still some scrolling oddness but I don't think that's fixable. :shipit: 

and-you-get-a-scrollbar

**Comment by justinshreve** [comment]
Fixed the second modal scrollbar. I don't think there is anything we can do about the background one.

**Pull request merged**